proxylib/kafka/parser.go

```diff
@@ -0,0 +1,206 @@
+	if !ok {
+		log.Warning("Matches() called with type other than KafkaRequest")
+		return false
+	}
+	if rule.Topic == "" || len(req.topics) == 0 {
+		return req.RuleMatches(&rule.PortRule)
+	} else if _, exists := req.topics[rule.Topic]; exists && req.RuleMatches(&rule.PortRule) {
+		delete(req.topics, rule.Topic)
+		return len(req.topics) == 0
+	}
+	return false
+}
+
+// KafkaRuleParser parses protobuf L7 rules to enforcement objects
+// May panic
+func KafkaRuleParser(rule *cilium.PortNetworkPolicyRule) []L7NetworkPolicyRule {
+	l7Rules := rule.GetL7Rules()
+	if l7Rules == nil {
+		return nil
+	}
+
+	allowRules := l7Rules.GetL7Rules()
+	rules := make([]L7NetworkPolicyRule, 0, len(allowRules))
+	for _, l7Rule := range allowRules {
+		var kr KafkaRule
+		for k, v := range l7Rule.Rule {
+			switch k {
+			case "role":
+				kr.Role = v
+			case "apikey":
+				kr.APIKey = v
+			case "apiversion":
+				kr.APIVersion = v
+			case "clientid":
+				kr.ClientID = v
+			case "topic":
+				kr.Topic = v
+			default:
+				ParseError(fmt.Sprintf("Unsupported key: %s", k), rule)
+			}
+		}
+		err := kr.PortRule.Sanitize()
+		if err != nil {
+			panic("invalid Kafka rule")
+		}
+		rules = append(rules, &kr)
+	}
+	return rules
+}
+
+type KafkaParserFactory struct{}
+
+var kafkaParserFactory *KafkaParserFactory
+
+func init() {
+	log.Info("init(): Registering kafkaParserFactory")
+	RegisterParserFactory(parserName, kafkaParserFactory)
+	RegisterL7RuleParser(parserName, KafkaRuleParser)
+}
+
+type KafkaParser struct {
+	connection *Connection
+}
+
+func (pf *KafkaParserFactory) Create(connection *Connection) interface{} {
+	p := KafkaParser{connection: connection}
+	return &p
+}
+
+func (p *KafkaParser) OnData(reply bool, reader *Reader) (OpType, int) {
+	length := reader.Length()
+	if length == 0 {
+		return NOP, 0
+	}
+
+	framelength := 4          // account for the length field
+	lenbuf := make([]byte, 8) // Peek the first eight bytes
+	n, err := reader.PeekFull(lenbuf)
+	if err == nil {
+		framelength += int(binary.BigEndian.Uint32(lenbuf[:4]))
+	} else {
+		// Need more data
+		return MORE, 8 - n
+	}
+
+	if reply {
+		// Replies are always passed as-is. No need to parse them
+		// on top of the frame length and correlation ID.
+		correlationID := binary.BigEndian.Uint32(lenbuf[4:])
+		p.connection.Log(cilium.EntryType_Response,
+			&cilium.LogEntry_GenericL7{
+				GenericL7: &cilium.L7LogEntry{
+					Proto: parserName,
+					Fields: map[string]string{
+						"CorrelationID": strconv.Itoa(int(correlationID)),
+					},
+				},
+			})
+		return PASS, framelength
+	}
+
+	// Ask for more if full frame has not been received yet
+	if length < framelength {
+		// Not enough data, ask for more and try again
+		return MORE, framelength - length
+	}
+
+	var req KafkaRequest
+	req.RequestMessage, err = kafka.ReadRequest(reader)
+	if err != nil {
+		if flowdebug.Enabled() {
+			log.WithError(err).Warning("Unable to parse Kafka request; closing Kafka connection")
```

Comment:
Does it print metadata (e.g., correlation ID) for the request in the log output?

Reply:
Did not, added. For now the errors use the same code as (policy) drops, will need to consider if that may cause confusion.